src/Compilers/CSharp/Portable/Lowering/Instrumentation/LocalStateTracingInstrumenter.cs

```diff
@@ -0,0 +1,621 @@
+    ///    </code>
+    ///
+    /// 2) <see cref="BoundStateMachineInstanceId"/>
+    ///    This node represents a reference to a synthesized state machine instance id field of the state machine. Lowered to a field read during state machine lowering.
+    ///
+    /// 3) <see cref="BoundParameterId"/>/<see cref="BoundLocalId"/>
+    ///    Represents id of a user-defined parameter/local. Emitted as ldc.i4 of either the parameter/local ordinal if the variable was not lifted, 
+    ///    or the token of its hoisted field.
+    ///
+    /// Each local variable write is followed by a call to one of the <c>LogLocalStoreXyz</c> or <c>LogParameterStoreXyz</c> instance methods on <c>$context</c>.
+    /// Writes to locals passed to a function call site by-ref are logged after the call returns.
+    /// <c>LogParameterStoreXyz</c> are emitted on explicit parameter assignment and also at the beginning of a method with parameters to log their initial values.
+    /// 
+    /// The loggers are specialized to handle all kinds of variable types efficiently (without boxing).
+    /// Specialized loggers are also used to track local variable aliases via ref assignments.
+    /// </remarks>
+    internal sealed class LocalStateTracingInstrumenter : CompoundInstrumenter
+    {
+        /// <summary>
+        /// Represents instrumentation scope - i.e. method, lambda or local function body.
+        /// We define a new <see cref="ContextVariable"/> (of LocalStoreTracker well-known type) in each scope,
+        /// so that this variable is always directly accessible within any instrumented code and always stack-allocated.
+        /// </summary>
+        private sealed class Scope
+        {
+            public LocalSymbol ContextVariable;
+            private ArrayBuilder<LocalSymbol>? _lazyPreviousContextVariables;
+
+            public Scope(LocalSymbol contextVariable)
+            {
+                ContextVariable = contextVariable;
+            }
+
+            public void Open(LocalSymbol local)
+            {
+                _lazyPreviousContextVariables ??= ArrayBuilder<LocalSymbol>.GetInstance();
+                _lazyPreviousContextVariables.Push(ContextVariable);
+                ContextVariable = local;
+            }
+
+            public void Close(bool isMethodBody)
+            {
+                if (_lazyPreviousContextVariables is { Count: > 0 })
+                {
+                    ContextVariable = _lazyPreviousContextVariables.Pop();
+                }
+
+                if (isMethodBody)
+                {
+                    Debug.Assert(_lazyPreviousContextVariables?.IsEmpty() != false);
+                    _lazyPreviousContextVariables?.Free();
+                }
+            }
+        }
+
+        private readonly Scope _scope;
+        private readonly SyntheticBoundNodeFactory _factory;
+        private readonly BindingDiagnosticBag _diagnostics;
+        private readonly TypeSymbol _instrumentationType;
```

Comment:
>_instrumentationType [](http://example.com/codeflow?start=36&length=20)

`_localStoreTrackerType`?

Reply:
Renamed to `_contextType` since we also have `ContextVariable` on the `Scope`.